python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1282,6 +1282,63 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ButlerURI:
                                "Use Dataastore.getURIs() instead.")
         return primary
 
+    def getArtifacts(self, refs: Iterable[DatasetRef],
+                     destination: ButlerURI, transfer: str = "auto",
+                     preserve_path: Optional[bool] = True) -> List[ButlerURI]:
+        """Retrieve the file artifacts associated with the supplied refs.
+
+        Parameters
+        ----------
+        refs : iterable of `DatasetRef`
+            The datasets for which file artifacts are to be retrieved.
+            A single ref can result in multiple files. The refs must
+            be resolved.
+        destination : `ButlerURI`
+            Location to write the file artifacts.
+        transfer : `str`, optional
+            Method to use to transfer the artifacts. Must be one of the options
+            supported by `ButlerURI.transfer_from()`.
+        preserve_path : `bool`, optional
+            If `True` the full path of the file artifact within the datastore
+            is preserved. If `False` the final file component of the path
+            is used.
+
+        Returns
+        -------
+        targets : `list` of `ButlerURI`
+            URIs of file artifacts in destination location. Order is not
+            preserved.
+        """
+        if not destination.isdir():
+            raise ValueError(f"Destination location must refer to a directory. Given {destination}")
+
+        # Source -> Destination
+        to_transfer: Dict[ButlerURI, ButlerURI] = {}
```

Comment:
This presumes that the set of artifacts to be transferred is relatively small.  You might want to consider a generator-based implementation in case there are millions of artifacts (but we are not likely to reach that scale for quite a while, if ever).

Reply:
I can do the transfer in the loop (allowing the refs from queryDatasets to go straight to transfer) but fundamentally I think the right answer is to use asyncio or something (especially for S3 to file). If I do that the command line tool won't be able to report the count of the number of refs either (which might be fine but I worried that that would leave the command line tool essentially blocking for minutes as the query happens and then the transfer from S3 happens).  In chained datastore it's a bit harder to save on memory because I have to filter before I forward the request to the child datastore.

Reply:
The dict does actually help here because queries can result in repeated DatasetRef and this dict has the side effect of filtering those out so that you don't try to copy the same file multiple times.